Add draft tutorial for jni programming (#251)
jni-rs/jni-rs: book/src/counter_solution.md

# Counter Solution

This section walks through one possible implementation of a Rust-backed Counter
for Java. The Java side is purposefully oversimplified, so that we may focus on
Rust (which would not change, with a more correct Java implementation).

## Java Counter API
First, we need to add a `close()` to the counter, so that we can free native
resources.

```java

> // Autocloseable is used to allow try-with-resources
> class Counter implements Autocloseable {

Reviewer: ```suggestion
// AutoCloseable is used to allow try-with-resources
class Counter implements AutoCloseable {
```

Reviewer: And below

ptr` field, which will be initialized on construction. `ptr` will then be passed
down to each of the native methods, so that the count may be fetched or
incremented, or dropped.

```java
// Mandatory disclaimer: In practice, you'll find that the Java side is usually
// much more involved than what is presented here, so that it can avoid
// double-frees, use-after-free, and guarantee cleanup. We intentionally put this
// off for now, because good solutions won't involve any Rust code, and they ramp
// up the difficulty.
class Counter implements Autocloseable {
    // The pointer to the counter in the native (Rust's) heap.
    private final long ptr;

    public Counter() {
        this.ptr = NativeAPI.counter_new();
    }

    public int get() {
        return NativeAPI.counter_get(ptr);
    }

    public int increment() {
        return NativeAPI.counter_increment(ptr);
    }

    public void close() {
        // Note the following problems with this implementation of close:
        //
        // 1. Close is not idempotent, as encouraged in the Autocloseable docs
        // 2. The counter may be double freed.
        // 3. Close may race with calls to get or increment

>         // 4. It's not guaranteed that a caller will call close.

Reviewer: I wouldn't say that both are necessarily problems:
(3) — not an issue unless the class is specified as thread-safe (i.e., it is a misuse to call increment and close from different threads).
(4) — depending on how the class is used, it might be a valid choice to make it `AutoCloseable` and document as one that must be closed by its users (e.g., using try-with-resources). In some cases, it will get unwieldly (many objects; or long-living objects), and other solutions would be better (`Cleaner`s or context-based approaches).

    _env: JNIEnv,
    _class: JClass,
) -> jlong {
    let boxed_counter = Box::new(Counter::new());
    Box::into_raw(boxed_counter) as jlong
}

#[no_mangle]
pub extern "system" fn Java_jni_1rs_1book_NativeAPI_counter_1increment(
    _env: JNIEnv,
    _class: JClass,
    ptr: jlong,
) -> jint {
    let counter = unsafe { &mut*(ptr as *mut Counter) };
    let counter_value = counter.increment();
    return counter_value;
}

#[no_mangle]
pub extern "system" fn Java_jni_1rs_1book_NativeAPI_counter_1get(
    _env: JNIEnv,
    _class: JClass,
    ptr: jlong,
) -> jint {
    let counter = unsafe { &*(ptr as *mut Counter) };
    let counter_value = counter.get();
    return counter_value;
}

#[no_mangle]
pub extern "system" fn Java_jni_1rs_1book_NativeAPI_counter_1destroy(
    _env: JNIEnv,
    _class: JClass,
    ptr: jlong,
) {
    unsafe { Box::from_raw(ptr as *mut Counter) };
}
```

## Summary
In this exercise, we examined a few ways to wrap a Rust Counter for use from
Java. You should be familiar with how panic can interact with `drop`, and some
approaches for preventing resources from being dropped prematurely. There are
still these problems with the counter:

> 1. `increment` and `get` may use the counter after its been freed.

Reviewer: ```suggestion
1. `increment` and `get` may use the counter after it's been freed.
```

2. it can be double freed ( via multiple `close` calls).
3. or perhaps never freed (never call `close`).
4. `close()` is not idempotent, as encouraged in the [`Autocloseable`
   docs](https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html#close--).

> 5. it's not threadsafe; it's possible to call `get` during `destroy`, for

Reviewer: ```suggestion
5. it's not thread-safe; it's possible to call `get` during `destroy`, for
```
